File: src/runtime/boxes/BbLen.cs
// BbLen.cs — LEN: match exactly n characters
// Mirrors src/runtime/boxes/bb_len.c
//
// α: if cursor+n > Ω → ω;  cursor += n; γ
// β: cursor -= n; ω

namespace Snobol4.Runtime.Boxes;

public sealed class BbLen : IByrdBox
{
    private readonly int _n;
    public BbLen(int n) { _n = n; }

    public Spec Alpha(MatchState ms)
    {
        if (ms.Cursor + _n > ms.Length)  return Spec.Fail;
        var result = Spec.Of(ms.Cursor, _n);
        ms.Cursor += _n;
        return result;
    }

    public Spec Beta(MatchState ms)
    {
        ms.Cursor -= _n;
        return Spec.Fail;
    }
}

File: src/runtime/boxes/BbLen.java
/**
 * BbLen.java — LEN: match exactly n characters
 * Port of bb_len.c / bb_len.s
 *
 *   LEN_α:  if (Δ+n > Ω)            goto LEN_ω;
 *           LEN=spec(Σ+Δ,n); Δ+=n;  goto LEN_γ;
 *   LEN_β:  Δ-=n;                   goto LEN_ω;
 */
class BbLen extends BbBox {
    private final int n;

    public BbLen(MatchState ms, int n) { super(ms); this.n=n; }

    @Override public Spec alpha() {
        if (ms.delta + n > ms.omega) return null;
        Spec r = new Spec(ms.delta, n);
        ms.delta += n;
        return r;
    }

    @Override public Spec beta() { ms.delta -= n; return null; }
}
